충남대 BE_이은경 2주차 과제 (3단계)

pkyung opened the pull request:
## 과제 진행
### 위시 리스트
- [x] 위시 리스트 도메인 설계
- [x] 위시 리스트 상품 추가
- [x] 위시 리스트 상품 목록 조회
- [x] 위시 리스트 상품 삭제

pkyung committed:
Chore: 1단계 코드 이동

pkyung committed:
Chore: 기능 목록 이동

pkyung committed:
Chore: 1주차 피드백 반영 코드 이동

pkyung committed:
Docs: 기능 목록 작성 - 1단계

pkyung committed:
Refactor: final 키워드를 추가하여 autowired 제거

pkyung committed:
Docs: 기능 목록 수정

pkyung committed:
Feat: ValidProductName 어노테이션 생성

ValidProductName 어노테이션을 생성하고 ProductNameValidator를 구현하여
상품 이름의 길이가 15자 이내인지, 특정 특수 문자만 들어있는지, 카카오
라는 단어가 들어있는 지 확인하고, 그에 맞는 응답을 하도록 구현했다.
controller에 ExceptionHandler를 통해 validated로 발생되는
MethodArgumentNotValidException 에러를 잡아서 내부의 defaultMessage를
가져와 400 bad request 와 함꼐 메시지를 반환하도록 구현했다.

pkyung committed:
Refactor: 필요 없는 import 제거 및 코드 정렬

pkyung committed:
Docs: 기능 목록 구현 - 1단계

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: 기본 Validate 어노테이션으로 수정함

Validator를 제작해서 처리했던 상품 이름 validation을 기본 Validate 어노테이션을 활용하여 구현함

pkyung committed:
Refactor: ExceptionHandler 클래스로 분리

컨트롤러에 예외처리 로직이 있어서 코드가 복잡해졌다. 이를
ProductExceptionHandler 클래스를 만들어 ProductController의 예외를
처리하도록 구현했다.

pkyung committed:
Refactor: Validated를 Valid로 변경

pkyung committed:
Docs: 기능 목록 작성 - 2단계

pkyung committed:
Refactor: 패키지 정리

pkyung committed:
Feat: 사용자 모델 설계

2단계 과제의 요청에 따라 user 객체에는 email과 password를 필드로
설정했다. 그리고 schema에 user 테이블을 생성하는 sql을 작성했다.

pkyung committed:
Feat: 사용자 회원 가입 api

1. userDao : insert와 findByEmail을 만들어 사용자를 데이터베이스에
   삽입하고, 조회하도록 구현했다.
2. UserService : makeUser와 getUser로 사용자 생성, 조회의 비즈니스
   로직을 처리한다. 이 때, 계징이 이미 존재하거나, 존재하지 않는 계정일
   때, 미리 만들어둔 예외를 던진다.
3. UserController : PostMapping과 GetMapping을 통해 사용자 요청을
   처리한다.
4. exception : UserAlreadyExistsException과 UserNotFoundException을
   생성했고, 이를 UserExceptionHandler에서 담당하여 controller에서
   에러 발생 시에 처리한다.
5. UserDto : client와 controller 사이, controller와 service 사이에서
   값을 전달한다.

pkyung committed:
Feat: 사용자 로그인 시에 token 발급

1. JwtUtil : 메서드를 모두 static으로 만들어 다른 클래스에서 의존성
   주입을 하지 않아도 된다. 이 클래스는 JWT 토큰을 생성하고 만료되었는지
   확인하는 등의 메서드가 구현되어 있다.
2. LoginService : 로그인 시, 사용자 정보가 있는 지, 비밀번호가 일치하는
   지 확인하고 맞다면 토큰을 만들어 반환한다. 없다면 사용자를 찾을 수
   없다는 에러를 반환한다. 이 에러는 미리 만들어 둔
   UserExceptionHandler에서 함꼐 잡는다.
3. LoginContorller : 로그인 요청이 들어오면 service를 호출한다.
4. build.gradle : jwt 관련 의존성을 추가했다.

pkyung committed:
Docs: 기능 목록 구현 - 2단계

pkyung committed:
Merge remote-tracking branch 'upstream/pkyung'

# Conflicts:
#	README.md
#	src/main/java/gift/controller/ProductController.java
#	src/main/java/gift/model/Product.java
#	src/main/java/gift/repository/ProductDao.java

pkyung committed:
Refactor: 2단계 멘토님 피드백 반영

1. dto에 request나 response가 들어있다면 dto로 당연시 인식되기 때문에
   dto를 생략 가능하다.
2. LoginController, LoginService를 따로 나누면 loginout 시에도
   LogoutController, LogoutService를 생성해야 하기 때문에
   AuthController, AuthService로 이름을 수정했다.
3. ProductDto의 @NotEmpty는 @Length와 겹쳐 제거했다.
4. JwtUtil의 secretKey는 중요한 정보이기 때문에 application.properties에
   저장한 값을 가져왔다. 하지만 static이기 때문에 Value 어노테이션으로
   한 번에 가져올 수는 없었고, setter를 통해 가져왔다.
5. expiredTime을 상수로 지정했다.

pkyung committed:
Fix: 비밀번호 확인 로직 반대로된 것을 확인 후 수정

pkyung committed:
Refactor: application.properties ignore 파일에 추가

pkyung committed:
Refactor: 2단계 요구사항 명세 수정

1. 기존의 user를 생성하는 makeUser 메서드를 Auth로 옮겨 register로 수정
2. accessToken이던 이름을 token으로 변경

pkyung committed:
Fix: 비밀번호 틀릴 시 forbidden 반환으로 수정

pkyung committed:
Refactor: exception 관리 패키지 별로 수정함

pkyung committed:
Refactor: 멘토님 피드백 반영

1. requestDto -> request 로 변수명 수정
2. LoginRequest와 UserReqeust 의 email validate 추가
3. email pk id로 auto increment
4. User 객체에 matchPassword 추가
5. 한 개만 조회 시에 optional 객체 사용

pkyung committed:
Refactor: 필요없는 구절 제거

pkyung committed:
Refactor: 멘토님 피드백 반영

1. matchPassword 수정
2. email unique 하게 수정

pkyung committed:
Docs: 기능 목록 작성 - 3단계

pkyung committed:
Feat: wish 도메인 설계

pkyung committed:
Feat: 위시 리스트 상품 추가

1. LoginUser 어노테이션 생성함
2. LoginUserArgumentResolver를 통해 Bearer 토큰이 아니거나, 비어있거나,
   만료되었다면 커스텀한  UnauthorizationException 에러 던짐 그렇지
   않다면 User 객체 반환
3. WishDao의 insert를 통해 Wish 삽입
4. WishService를 통해 ProductId의 상품이 존재하는지 확인 후 wishdao에
   삽입
5. WishController에서 Post 요청과 함께 토큰이 함께 오면 service에서
   처리시킴

pkyung committed:
Feat: 위시 리스트 상품 목록 조회

1. wishes 테이블의 productId 필드를 product의 id 와 외래키로 맺어줌
2. WishDao에서 join query를 통해 상품 리스트 조회
3. WishService, WishController를 구현함

pkyung committed:
Feat: 위시 리스트 상품 삭제

1. productId로 삭제할 상품 id를 받음
2. productId가 본인의 위시 리스트에 없을 시 exception 반환함
3. Service와 Controller를 구현함

pkyung committed:
Docs: 기능 목록 구현 - 3단계

pkyung committed:
Merge branch 'pkyung' into main

ksy90101 commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
```
👍

ksy90101 commented on src/main/java/gift/dto/WishRequest.java:
```diff
@@ -0,0 +1,9 @@
+package gift.dto;
+
+public class WishRequest {
+    private Long productId;
```
여기도 validate를 해보면 좋을거 같아요~
습관이 중요한거 같습니다!

ksy90101 replied:
여기 리뷰 반영안되었습니다!

ksy90101 commented on src/main/java/gift/exception/wish/WishExceptionHandler.java:
```diff
@@ -0,0 +1,23 @@
+@RestControllerAdvice(assignableTypes = WishController.class)
+public class WishExceptionHandler {
+
+    @ExceptionHandler(ProductNotFoundException.class)
+    public ResponseEntity<String> handleProductNotFoundException(ProductNotFoundException e) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
+    }
```
이제 슬슬 중복이 생기는거 같아요.
이렇게 모든 모델마다 NotFoundException을 만든다면 계속 중복코드가 생기지 않을까요?

ksy90101 commented on src/main/java/gift/exception/wish/WishExceptionHandler.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @ExceptionHandler(UnauthorizedException.class)
+    public ResponseEntity<String> handleUnauthorizedException(UnauthorizedException e) {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(e.getMessage());
+    }
```
이것도 중복같아요
혹시 이런경우 어떻게 해결할수 있을지 고민해보면 어떨까요?
이건 좀 범용적인 예외가 아닐까요?

ksy90101 commented on src/main/java/gift/model/Wish.java:
```diff
@@ -0,0 +1,26 @@
+public class Wish {
+    private Long id;
+    private Long productId;
+    private String userEmail;
```
왜 userEmail을 넣어두었나요~?
user_id를 넣어도 되지 않을까요?

ksy90101 commented on src/main/java/gift/repository/WishDao.java:
```diff
@@ -0,0 +1,51 @@
+        String sql = "SELECT p.id, p.name, p.price, p.imageUrl " +
+                "FROM products p " +
+                "JOIN wishes w ON p.id = w.productId " +
+                "WHERE w.userEmail = ?";
```
이렇게 되었을때 user_id가 있는 의미가 별로 없을거 같아요!

ksy90101 commented on src/main/java/gift/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,44 @@
+            throw new UnauthorizedException("인증되지 않은 사용자입니다.");
+        }
+
+        String token = authorizationHeader.substring(7);
```
7이라는 단어는 무엇인지 저는 알기가 어려운데요
어떤 의미로 썼는지 상수로 나타내면 어떨까요?

ksy90101 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,42 @@
+import gift.repository.WishDao;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+
+@Service
+public class WishService {
```
```suggestion
import java.util.List;

@Service
public class WishService {
```

개행 하나도 의미전달의 의미가 있습니다. 이러한 부분들을 생각해보시면 좋을거 같아요~

ksy90101 commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,42 @@
+        Wish wish = new Wish(request.getProductId(), email);
+        wishDao.insert(wish);
+        return wish;
+    }
+    public List<Product> getAllWishProductsByUser(String email) {
+        return wishDao.findAll(email);
+    }
```
```suggestion
    }
    
    public List<Product> getAllWishProductsByUser(String email) {
        return wishDao.findAll(email);
    }
```

ksy90101 commented on src/main/java/gift/controller/WishController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @PostMapping("/wishes")
+    public ResponseEntity makeWish(@RequestBody WishRequest request, @LoginUser User user) {
```
user를 넘기는 방법도 있엇을텐데, email만 넘긴 이유가 있을까요!?

ksy90101 reviewed (changes requested):
빠르게 3단계 해주셨네요!
고생 많으셨습니다~
간단한 피드백 남겼으니 한번 확인해주세요~
언제든 궁금한게 있다면 DM주세요!

pkyung committed:
Refactor: 멘토님 피드백 반영

1. 중복된 exceptionHandler를 global로 분리하여 통합시킴
2. userEmail로 처리하던 것을 userId로 처리함
3. 코드 개행 수정

ksy90101 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
제가 이야기한것이 하나로 다 합치라는 의미는 아니였어요..
공통 예외는 공통으로.

각 컨트롤러의 예외는 각 컨트롤러 핸들러에 넣으면 어떨까 라는 생각이였습니다 ㅎㅎ

pkyung replied:
생각해보니 UserException이나 ProductException에서 같은 status를 반환할 것 같아서 합친 것인데 혹시 어떤 기준으로 공통이라는 것을 찾아야 하나요??

ksy90101 replied:
지금 처럼 해도 되고, 각 컨트롤러에만 사용하는 에러는 해당 컨트롤러의 Handler를 통해 응집도를 높혀보자는 의미였어요.

NotFound같은 경우엔 거의다 로직이 비슷할거 같아요.

그런 경우 NotFoundException을 만들고 처리하긴 합니다 ㅎㅎ

굳이 예외를 다 만들필요가 없다고 생각이 들어서요!

그런 의미였어요~

ksy90101 commented on src/main/java/gift/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,45 @@
+            throw new UnauthorizedException("인증되지 않은 사용자입니다.");
+        }
+
+        String token = authorizationHeader.substring(BEARER_LENGTH);
```
7이 그런 의미라면

"Bearer "를 상수화하고, length를 쓰면 어떨까요?
위의 Bearer이 변경되면 또 여기도 신경ㄹ써야 하는데 한번에 변경되지 않을까요?

ksy90101 commented on src/main/java/gift/model/LoginUser.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface LoginUser {
```
👍

ksy90101 commented on src/main/java/gift/dto/WishRequest.java:
```diff
@@ -0,0 +1,9 @@
+package gift.dto;
+
+public class WishRequest {
```
좀더 심화적인 내용으로 Java17을 쓰니 record를 공부해서 한번 적용해보면 어떨까요?

ksy90101 reviewed (changes requested):
리뷰 반영 고생하셨어요~
간단한 피드백을 남겼습니다!
마지막으로 한번만 더 확인해보시죠!

pkyung committed:
Refactor: 멘토님 피드백 반영

1. dto를 record로 변경함 -> getter나 생성자, 필드 캡슐화, equals 메서드,
   hashCode 메서드, toString 메서드를 구현하지 않아도 자동으로
   생성되어 코드가 깔끔해지고 편리해, 대신 get으로 사용하던 것을
   필드이름() 으로 변경해야 했음
2. WishRequest의 productId는 auto_increment에 의해서 최소 1부터 시작되니
   Min(1)을 통해 valid 판단을 함
3. Bearer를 상수화하여 사용함

ksy90101 commented on src/main/java/gift/controller/UserController.java:
```diff
@@ -16,7 +16,13 @@ public UserController(UserService userService) {
+    }
+
+    @GetMapping("/users/{id}")
+    public ResponseEntity<User> getUserById(@PathVariable("id") Long id) {
```
PathVariable의 value인 id는 위의 이름과 동일하다면 없애도 될거 같습니다~

ksy90101 reviewed (approved):
고생하셨습니다~
이만 머지할게요!
언제든 궁금한게 있다면 DM주세요!
